tests/recordCallsToSetTimeout.ts

```diff
@@ -0,0 +1,193 @@
+   * call.
+   * @param duration - The duration associated with a particular `setTimeout`
+   * call.
+   * @returns An instance of NodeJS.Timeout which is only supplied to fulfill
+   * the existing type of `setTimeout` and serves no purpose.
+   */
+  _mockSetTimeoutImplementation = (
+    callback: SetTimeoutCallback,
+    duration: number | undefined = 0,
+  ): NodeJS.Timeout => {
+    // Build a "fake" Timeout object
+    const timeout = this.originalSetTimeout(EMPTY_FUNCTION, 0);
+    const interceptedCallback = this.interceptCallback(
+      callback,
+      this._stopPassingThroughCalls,
+    );
+    const call = {
+      callback: interceptedCallback,
+      duration,
+      timeout,
+    };
+    this.calls.push(call);
+
+    if (this.numAutomaticCallsRemaining > 0) {
+      call.callback();
+      this.numAutomaticCallsRemaining -= 1;
+    } else {
+      this.events.emit('numCallsToPassThroughExhausted');
+    }
+    this.events.emit('setTimeoutAdded');
+    return timeout;
+  };
+
+  /**
+   * The function with which to replace the global `clearTimeout` function. This
+   * mock implementation will find a call to `setTimeout` that returned the
+   * given Timeout object and remove it from the queue. If no such call has been
+   * made, then this does nothing.
+   *
+   * @param timeout - A Timeout object as returned by `setTimeout`.
+   */
+  _mockClearTimeoutImplementation = (timeout: NodeJS.Timeout): void => {
+    const index = this.calls.findIndex((c) => c.timeout === timeout);
+
+    if (index !== -1) {
+      this.calls.splice(index, 1);
+    }
+  };
+
+  _stopPassingThroughCalls = () => {
+    this.numAutomaticCallsRemaining = 0;
+  };
+}
+
+/**
+ * Replaces the global `setTimeout` function with one which, upon being called,
+ * records the callback given to it. The callback may be stored in a queue to be
+ * called later using `next` / `nextMatchingDuration`, or it may be called
+ * immediately.
+ *
+ * @param options - The options.
+ * @param options.numAutomaticCalls - By default, it is up to you to manually
+ * call `setTimeout`s that have been queued. If you know the number of times
+ * `setTimeout` should be called within a test, however, you may specify that
+ * here, and each time `setTimeout` is called, its callback will be called
+ * immediately, up to this many times (default: 0).
+ * @param options.interceptCallback - A function that can be used to replace a
+ * callback that is passed to `setTimeout`, allowing you to call it yourself
+ * (perhaps in a `try`/`catch` block, or something else).
+ * @returns An object that can be used to interact with calls to `setTimeout`.
+ */
+export default function recordCallsToSetTimeout({
+  numAutomaticCalls = 0,
+  interceptCallback = (callback) => callback,
+}: {
+  numAutomaticCalls?: number;
+  interceptCallback?: InterceptingCallback;
+} = {}): SetTimeoutRecorder {
+  const originalSetTimeout = global.setTimeout;
```

Comment:
Nit: Any reason this is passed in, rather than being used directly by the `SetTimeoutRecorder` constructor?

Reply:
Because by the time that SetTimeoutRecorder is called, `setTimeout` is already stubbed. So we need to get access to the original version first before it's stubbed, then pass it into SetTimeoutRecorder.

That said, this doesn't need to be passed in at all, it can just be a global at the top of the file. This is changed in [46b50aa](https://github.com/MetaMask/eth-block-tracker/pull/103/commits/46b50aa1e8caecced29a6b2e4c7b17b71471083e).